### PR TITLE
:bug: KCP: Allow unsetting fields if their child fields are mutable

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
This PR makes parent fields mutable if there fields already are.

An example is that it was possible to modify `rolloutBefore.certificatesExpiryDays` but not to unset `rolloutBefore`.

I did this for all fields as I think we want to allow to unset all of them. In some cases it's theoretically not necessary because when the field is unset either it's marshalled as an empty object or defaulting is run. But I think it's easier to make them all mutable instead of documenting the various cases.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Review by killianmuldoon — APPROVED** [review]
/lgtm

**Review by ykakarap — COMMENTED** [review]
/lgtm

Do we also need to cherry-pick this back to release-1.2. release-1.2 does not have the `rolloutBefore` field but does it still apply for the other fields?

**Comment by sbueringer** [comment]
Good idea. Let's do it. Don't want to deal with this problem again

**Comment by sbueringer** [comment]
(just to get notified)

File: controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go [hunk 1]
```diff
@@ -141,16 +141,22 @@ func (in *KubeadmControlPlane) ValidateUpdate(old runtime.Object) error {
 		{"metadata", "*"},
 		{spec, kubeadmConfigSpec, clusterConfiguration, "etcd", "local", "imageRepository"},
 		{spec, kubeadmConfigSpec, clusterConfiguration, "etcd", "local", "imageTag"},
+		{spec, kubeadmConfigSpec, clusterConfiguration, "etcd", "local", "extraArgs"},
 		{spec, kubeadmConfigSpec, clusterConfiguration, "etcd", "local", "extraArgs", "*"},
 		{spec, kubeadmConfigSpec, clusterConfiguration, "dns", "imageRepository"},
 		{spec, kubeadmConfigSpec, clusterConfiguration, "dns", "imageTag"},
 		{spec, kubeadmConfigSpec, clusterConfiguration, "imageRepository"},
+		{spec, kubeadmConfigSpec, clusterConfiguration, apiServer},
```
Comment (fabriziopandini) [inline controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go:149]: am I right is assuming that whenever we have somepath + "*" we always want to allow also somepath?
if yes, can this be resolved down in the code where we check allowed paths so we don't have to care about paying attention to apply the rule above everywhere?

Comment (sbueringer) [reply controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go:149]: We could, but I thought it's too magical if we have <parent> * in the list but implicitly also add <parent>

**Review by killianmuldoon — COMMENTED** [review]
/rename bug KCP: Allow unsetting fields if their child fields are mutable

**Comment by killianmuldoon** [comment]
/retitle  :bug: KCP: Allow unsetting fields if their child fields are mutable

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**